x-pack/plugins/security/server/lib/authentication/providers/token.js

```diff
@@ -356,7 +354,30 @@ export class TokenAuthenticationProvider {
       // it's called with this request once again down the line (e.g. in the next authentication provider).
       delete request.headers.authorization;
 
+      // If refresh fails with `400` then refresh token is no longer valid and we should clear session
+      // and redirect user to the login page to re-authenticate.
+      if (getErrorStatusCode(err) === 400 && canRedirectRequest(request)) {
+        this._options.log(
+          ['debug', 'security', 'token'],
+          'Clearing session since both access and refresh tokens are expired.'
+        );
+
+        // Set state to `null` to let `Authenticator` know that we want to clear current session.
+        return AuthenticationResult.redirectTo(this._getLoginPageURL(request), null);
+      }
+
       return AuthenticationResult.failed(err);
```

Comment:
__note:__ I kept the same behavior for AJAX requests that was before and similar to what we have for `saml`. If we decide we want to treat AJAX requests differently it'd make sense to do it separately for both `token` and `saml`.